### DISCUSSION TITLE
[@types/yargs] Regression with the "command" method

### DISCUSSION THREAD

**Discussion by astorije** [DefinitelyTyped/DefinitelyTyped#58457] [Issues with a @types package]
Consider the following code:

```ts
import yargs from "yargs";

function foo({ bar, baz }: { bar: string; baz?: boolean }) {
  console.log("foo with bar =", bar, ", baz =", baz);
}

void yargs
  .scriptName("cli")
  .command(
    "foo",
    "Description",
    () =>
      yargs
        .positional("bar", { describe: "bar", type: "string" })
        .option("baz", { describe: "baz", type: "boolean" }),
    foo
  ).argv;
```

Starting with `@types/yargs` v17.0.8, this produces the following errors when running `tsc --build`:

```
index.ts:12:5 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '() => yargs.Argv<{ bar: string | undefined; } & { baz: boolean | undefined; }>' is not assignable to parameter of type '{ [key: string]: Options; }'.
      Index signature for type 'string' is missing in type '() => Argv<{ bar: string | undefined; } & { baz: boolean | undefined; }>'.

12     () =>
       ~~~~~
13       yargs
   ~~~~~~~~~~~
14         .positional("bar", { describe: "bar", type: "string" })
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15         .option("baz", { describe: "baz", type: "boolean" }),
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/@types/yargs/index.d.ts:155:9
    155         command<O extends { [key: string]: Options }>(
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    156             command: string | ReadonlyArray<string>,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
    161             deprecated?: boolean | string,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    162         ): Argv<T>;
        ~~~~~~~~~~~~~~~~~~~
    The last overload is declared here.


Found 1 error.
```

This builds fine with `@types/yargs` v17.0.8.
I believe something was introduced in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50722 that causes this.
/Cc @forivall @bcoe @peterblazejewicz

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
I'm not an user. Isn't the 'bar' optional (`bar?`)? (non required)? so the arguments are not matched:

```ts
async function foo({ bar }: { bar?: boolean }) {
    console.log('foo with bar =', bar);
}

void yargs
    .scriptName('cli')
    .command(
            'foo',
            'Description',
            () => yargs.option('bar', { describe: 'bar', type: 'boolean' }),
            foo
        )
        .argv;
```

**Reply by astorije** [reply]
You're absolutely right, I updated the snippet above, thanks! However this does not fix the issue, the same error remains.

**Reply by peterblazejewicz** [reply]
the problem is, I didn't get the error. Note that my fat arrow method returns option, your original code returns empty object

**Reply by forivall** [reply]
it also might depend on which typescript version. we might need to bump the required version for the mapped camelcase types if there's a bug with that in some versions of typescript.

@astorije which typescript version are you using?

**Reply by astorije** [reply]
> Note that my fat arrow method returns option, your original code returns empty object

I didn't notice that, thanks for pointing it out! Unfortunately on my end (the code where this issue appeared) that error still shows up even with a returned value. The contrived example above isn't sufficient to highlight the issue it seems, let me try to reproduce it with a returned value.

> @astorije which typescript version are you using?

We're using v4.5.5, i.e. latest.

**Reply by astorije** [reply]
@peterblazejewicz, I have updated the snippet in the initial message (and the error message, similar but more descriptive) to which I hope is closer to the real-life failure. Let me know if you can reproduce this.

Thanks again for helping with this!
